lib/rules/no-misleading-character-class.js

```diff
@@ -181,6 +197,64 @@ module.exports = {
         const sourceCode = context.sourceCode;
         const parser = new RegExpParser();
 
+        /**
+         * Determines what offset to start a granular report, if it's straightforward to do so.
+         * @param {Character} char Individual character being reported on.
+         * @param {string} kind What kind of character report is being made.
+         * @param {Node} node Parent string node to report within.
+         * @param {Object} previousReport Kind and start offset for the previous report, if there was one.
+         * @returns {Object | null} Report offset, or null to report on the whole node.
+         * @see https://github.com/eslint/eslint/pull/17515#issuecomment-1707697186
+         */
+        function getReportStartOffset(char, kind, node, previousReport = {}) {
+            if (node.type !== "Literal" && (node.type !== "TemplateLiteral" || node.expressions.length)) {
+                return null;
+            }
+
+            // If a problematic character exists twice in the string, start each Nth search after N - 1
+            const searchStart = previousReport && previousReport.kind === kind
+                ? (previousReport.startOffset + 1)
+                : 0;
+
+            const searchableLength = searchStart + char.start + char.raw.length * 2;
+            const offsetStart = node.raw.slice(0, searchableLength).indexOf(char.raw, searchStart);
+
+            return offsetStart === -1 ? null : offsetStart;
+        }
+
+        /**
+         * Generates a granular loc for context.report.
+         * @param {Character} char Individual character being reported on.
+         * @param {string} kind What kind of character report is being made.
+         * @param {Node} node Parent string node to report within.
+         * @param {number} reportStartOffset Starting offset to report within the node.
+         * @returns {Object} Granular loc for context.report.
+         * @see https://github.com/eslint/eslint/pull/17515#issuecomment-1707697186
+         */
+        function generateReportLocation(char, kind, node, reportStartOffset) {
+            const reportEndOffset = reportStartOffset + char.raw.length;
+            let reportStartOffsetAdjusted = reportStartOffset;
+
+            if (!char.raw.startsWith("\\u")) {
+                reportStartOffsetAdjusted -= kindsForSingleCharacterAdjustment.has(kind) ? 1 : 2;
+            }
+
+            if (node.raw.slice(char.start + 1, char.start + 3) === "\\\\") {
```

Comment:
```suggestion
            } else if (node.type !== "Literal" || !node.regex) {
```

A problem is that `char.start` cannot be used to index `node.raw` because it does not represent an index in the source code of the containing node, but (probably) in the pattern of the generated regular expression. The way I've changed the logic fixes the off-by-one issue of the tests I've checked by assuming that a single backslash in the pattern will be matched by two backslashes in the source code if the containing node is not a regexp literal. There could be more edge cases I haven't considered though 😕